library/c/sandesh.h

```diff
@@ -25,6 +24,54 @@ extern "C" {
 #include <Wsk.h>
 #include <Ws2ipdef.h>
 
+#define SxExtAllocationSandeshTag 'SAND'
+
+static void *
+s_malloc(unsigned int size)
+{
+    void *mem = ExAllocatePoolWithTag(NonPagedPool, size+sizeof(unsigned int), SxExtAllocationSandeshTag);
+    *(unsigned int*)mem = size;
+    mem = (char*)mem + sizeof(unsigned int);
+    return mem;
+}
+
+static void *
+s_zalloc(unsigned int size)
+{
+    void *mem = s_malloc(size);
+    RtlZeroMemory(mem, size);
+
+    return mem;
+}
+
+static void
+s_free(void *mem)
+{
+    if (mem) {
+        mem = (char*)mem - sizeof(unsigned int);
+        ExFreePoolWithTag(mem, SxExtAllocationSandeshTag);
+    }
+}
+
+static void *
+s_realloc(void *mem, unsigned int size)
+{
+    unsigned int old_size;
+
+    void *mem_tmp = s_malloc(size);
+    mem = (char*)mem - sizeof(unsigned int);
+    old_size = *(unsigned int*)mem;
+    RtlCopyMemory(mem_tmp, mem, old_size);
+    s_free(mem);
```

Comment:
You have modified `mem` variable before calling `s_free`. It might cause a bug, because `s_free` also modifies it.